cmd/kube-proxy/app/server.go

```diff
@@ -392,25 +398,31 @@ type ProxyServer struct {
 
 // createClients creates a kube client and an event client from the given config and masterOverride.
 // TODO remove masterOverride when CLI flags are removed.
-func createClients(config componentconfig.ClientConnectionConfiguration, masterOverride string) (clientset.Interface, v1core.EventsGetter, error) {
-	if len(config.KubeConfigFile) == 0 && len(masterOverride) == 0 {
-		glog.Warningf("Neither --kubeconfig nor --master was specified. Using default API client. This might not work.")
-	}
+func createClients(config *componentconfig.KubeProxyConfiguration, masterOverride string) (clientset.Interface, v1core.EventsGetter, error) {
+	var kubeConfig *rest.Config
+	var err error
 
-	// This creates a client, first loading any specified kubeconfig
-	// file, and then overriding the Master flag, if non-empty.
-	kubeConfig, err := clientcmd.NewNonInteractiveDeferredLoadingClientConfig(
-		&clientcmd.ClientConfigLoadingRules{ExplicitPath: config.KubeConfigFile},
-		&clientcmd.ConfigOverrides{ClusterInfo: clientcmdapi.Cluster{Server: masterOverride}}).ClientConfig()
-	if err != nil {
-		return nil, nil, err
+	if config.UseServiceAccount {
```

Comment:
@liggitt @deads2k what if we just did this instead?
```go
loader := clientcmd.NewDefaultClientConfigLoadingRules()
loader.ExplicitPath = config.KubeConfigFile
kubeConfig, err := clientcmd.BuildConfigFromKubeconfigGetter(masterOverride, loader.Load)
```

I believe this would allow us to skip adding a "use service account" flag:
1. Use the kubeconfig file, if it's specified in the kube-proxy config
1. Support the master url override
1. Use the in-cluster config if the kubeconfig file is not specified

WDYT? If this is too big or risky of a change, then we can certainly keep what's here in this PR

Reply:
In fact, seems like `BuildConfigFromFlags()` has supported this use case?
https://github.com/kubernetes/kubernetes/blob/4ee72eb300423772020dd1cf208159058ba7dab5/staging/src/k8s.io/client-go/tools/clientcmd/client_config.go#L522-L539

If using InClusterConfig() when neither kubeconfig nor masterUrl is specified for kube-proxy is not considered breaking backwards-compat, I'd happy to do it this way. The current behavior is using default API client. cc @luxas

Reply:
Took another look, using `BuildConfigFromFlags()` seems valid here. As when none of `masterUrl` nor `kubeconfigPath` is defined, it will call `InClusterConfig()`. Then when `InClusterConfig()` fails, it then falls back to default config.  This seems to match the current behavior.

Reply:
> loader := clientcmd.NewDefaultClientConfigLoadingRules()

That would make `kube-proxy` honor `$KUBECONFIG` and `$HOME/.kube/config`, which seems problematic. Traditionally, we've had server components use explicit flags to specify connection config files, rather than picking up env-based config like $KUBECONFIG. 

`InClusterConfig()` falls somewhere between an explicit `--kubeconfig` and an implicit `$KUBECONFIG` or homedir behavior, since it is using paths and envs only intended to be set in a containerized environment.

Reply:
The kubeconfig is going to be specific in the config file via `--config`. We won't support `--kubeconfig` any more.

Maybe if the kubeconfig in the config file is blank, we fall back to in cluster?

Reply:
specific file falling back to InClusterConfig() if unspecified seems ok. Just make sure we don't pull in $KUBECONFIG or $HOME/.kube/config

Reply:
K

Reply:
@MrHohn can you make the change such that if `config.KubeConfigFile` is empty, we use the in-cluster config?

Reply:
@ncdc Sounds good, will ping again when ready.